### PR TITLE
xPack GNU Arm toolchain

### LINKED ISSUE #971: Debugger (arm-none-eabi-gdb) not working on up-to-date linux systems
Our [installation guide](https://modm.io/guide/installation/) suggests to install the ["official" precompiled GNU toolchain from ARM](https://developer.arm.com/Tools%20and%20Software/GNU%20Toolchain).
If you follow the instructions you will not have a working GDB debugger on modern Linux systems (e.g. Ubuntu ≥22.04 or Fedora ≥36) because ARM apparently requires Python3.8 (which is end-of-life since May 2021).
Also in our docker image(s) `arm-none-eabi-gdb` does not work for the same reason:

```
> docker run --rm -it ghcr.io/modm-ext/modm-build-cortex-m:latest
root@d8fbff209b23:/work# arm-none-eabi-gdb
Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Python path configuration:
  PYTHONHOME = (not set)
  PYTHONPATH = (not set)
  program name = '/usr/local/bld-tools/bld-tools-virtual-env/bin/python'
  isolated = 0
  environment = 1
  user site = 1
  import site = 1
  sys._base_executable = '/usr/local/bld-tools/bld-tools-virtual-env/bin/python'
  sys.base_prefix = '/usr'
  sys.base_exec_prefix = '/usr'
  sys.executable = '/usr/local/bld-tools/bld-tools-virtual-env/bin/python'
  sys.prefix = '/usr'
  sys.exec_prefix = '/usr'
  sys.path = [
    '/usr/lib/python38.zip',
    '/usr/lib/python3.8',
    '/usr/lib/lib-dynload',
  ]
Fatal Python error: init_fs_encoding: failed to get the Python codec of the filesystem encoding
Python runtime state: core initialized
ModuleNotFoundError: No module named 'encodings'

Current thread 0x00007f9e995ffc00 (most recent call first):
<no Python frame>
```
(Setting `PYTHONHOME`/`PYTHONPATH` does not help because python3.8 is compiled into the `arm-none-eabi-gdb` binary.)

### PR DESCRIPTION
**Pull request by rleh**
- [x] [docs] Installation guide: Use xPack GNU Arm toolchain instead Arm precompiled
  - [x] Linux
  - [ ] macOS
  - [x] Windows
- [x] [ci] Update docker images to `:2023-03-12` version
- [x] Add xPack formula to [modm-ext/homebrew-modm](https://github.com/modm-ext/homebrew-modm): modm-ext/homebrew-modm/pull/1
- [x] Diff the disassembly of a few interesting example
- [x] Test new compiler in hardware
  - [x] Tested by @rleh with STM32U575 and STM32F429
- [x] Don't strip extra compiler version info from [`MODM_BUILD_COMPILER`](https://modm.io/reference/module/modm-build/#build-information)

Fixes #971

### CODE REVIEW

**Review by salkinium — COMMENTED** [review]
(We also need to upgrade the macOS and Windows instructions, I'm not keen on supporting two toolchains).

**Comment by rleh** [comment]
> We also need to upgrade the macOS

Currently on MacOS the instructions are:
```sh
brew tap osx-cross/arm
brew install arm-gcc-bin@12
```
If you consider it better to use the xPack toolchain instead of Homebrew (I can't judge that for macOS), feel free to change that.

---

> and Windows instructions

Windows instructions are already updated for the xPack toolchain.

**Comment by salkinium** [comment]
> If you consider it better to use the xPack toolchain instead of Homebrew (I can't judge that for macOS), feel free to change that.

Yes, I'll add it to our homebrew-modm repo, just not today, cos I'm busy patching ST's HTML reference manuals again 🫠

**Comment by salkinium** [comment]
~~OpenOCD has a v0.12 release, so let's update that too while we're here. (I think it's only the instructions that need updating?)~~
Never mind, you already did.

File: docs/src/guide/installation.md
Comment (salkinium) [inline docs/src/guide/installation.md]:
```suggestion
- ARM toolchain: [toolchain-arm-xpack][] and [OpenOCD][] (at least v0.12!).
```

File: docs/src/guide/installation.md
Comment (salkinium) [inline docs/src/guide/installation.md]:
```suggestion
    Make sure to get at least OpenOCD release v0.12, since v0.11 is too old for
```

**Comment by rleh** [comment]
The MacOS jobs fails (that was expected) because the xpack toolchain formula in our homebrew repo is not yet merged (modm-ext/homebrew-modm#1).

File: docs/src/guide/installation.md
Comment (salkinium) [inline docs/src/guide/installation.md:86]: No? That's like the opposite of what we recommend 😜

File: docs/src/guide/installation.md
Comment (salkinium) [inline docs/src/guide/installation.md:38]: I see, the lib's name is getting ever more cursed, as it should be.

File: docs/src/guide/installation.md
Comment (rleh) [reply docs/src/guide/installation.md:86]: On linux systems with up-to-date packages (like Fedora or Arch) I recommend everyone to install the [distributions package(s)](https://packages.fedoraproject.org/pkgs/arm-none-eabi-gcc-cs/arm-none-eabi-gcc-cs-c++/) because it is so much less trouble (dependency management, some libs in wrong versions, $PATH, SElinux permissions, ...) than unpacking a tar archive to `/opt`.
(Only) the Ubuntu (22.04) package was broken in the past and is currently not up to date.

Comment (salkinium) [reply docs/src/guide/installation.md:86]: Yeah ok, I see the problem. Let's just try it and see if anything breaks. I feel like arm-none-eabi-gcc is more popular now, and so I guess there are more people who care about it.

**Comment by rleh** [comment]
Over at @roboterclubaachen the CI throws a lot of this error messages:
```
/opt/xpack-arm-none-eabi-gcc-12.2.1-1.2/bin/../lib/gcc/arm-none-eabi/12.2.1/../../../../arm-none-eabi/bin/ld: warning: build/lbuild-deps/scons-release/lbuild-deps.elf has a LOAD segment with RWX permissions
```

Probably related to their (faulty) code/linkerscript, but I would like to investigate before merging this PR.

cc @se-bi

**Comment by salkinium** [comment]
Might also be [the compiler version not getting parsed correctly](https://github.com/modm-io/modm/blob/develop/tools/build_script_generator/scons/resources/SConscript.in#L87) anymore.

**Comment by rleh** [comment]
> Might also be [the compiler version not getting parsed correctly](https://github.com/modm-io/modm/blob/develop/tools/build_script_generator/scons/resources/SConscript.in#L87) anymore.


With `arm-none-eabi-g++ (xPack GNU Arm Embedded GCC x86_64) 12.2.1 20221205`:
```python
>>> print(env.CompilerVersion())
120201
```

With `arm-none-eabi-g++ (Fedora 12.2.0-1.fc36) 12.2.0`:
```python
>>> print(env.CompilerVersion())
120200
```

➡️ `env.CompilerVersion()` works fine

**Comment by rleh** [comment]
> Over at @roboterclubaachen the CI throws a lot of this error messages:  `[...] has a LOAD segment with RWX permissions` [...]

The problem was only because old modm versions were used there, which I had overlooked.

**Comment by rleh** [comment]
@salkinium Have you had time to look into the MacOS GCC issue?

**Comment by salkinium** [comment]
No, sorry, I have to finish writing my paper until 1st May, then I have all the time.

**Comment by rleh** [comment]
> - [x]  Diff the disassembly of a few interesting example

I looked at some small STM32 examples and at the Nucleo-F429ZI unittest runner (which includes all available modules): There are only some (internal) lib names/prefixes changed, some code is ordered differently and thus a lot of addresses have changed.

---

@salkinium @chris-durand Are you ok with merging this or do you wan't to wait until we (you :D) have completely debugged the MacOS issue?

**Comment by salkinium** [comment]
The problem is that the macOS requires signed binaries, otherwise you need to manually whitelist dozens of dylibs, which is impractical. Homebrew signs the binaries ad-hoc, but breaks the Python installation.
So installation of this toolchain on macOS is just… broken in practice.
So please don't merge it yet. I need to fix this problem anyways for work, so it'll be fixed eventually I promise.

**Comment by rleh** [comment]
I would like to have the issue with the linux toolchain distribution (no functional debugger) fixed before next month and I consider it necessary to link to a functional toolchain distribution before we force all users to update their toolchains by merging #968.

I removed the MacOS related commit (0b8b9ccbf155ac5d36f08b1ce3238bea23212b89) from this branch and would like to merge this now. We can then deal with the MacOS stuff later.

**Review by salkinium — APPROVED** [review]
That's fair. We'll update the Windows and macOS CI later.

**Pull request merged**
